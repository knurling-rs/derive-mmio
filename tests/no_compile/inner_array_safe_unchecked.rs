mod inner {
    #[derive(derive_mmio::Mmio)]
    #[repr(C)]
    pub struct UartBank {
        data: u32,
    }

    impl UartBank {
        pub fn fake() -> UartBank {
            UartBank {
                data: 0x2,
            }
        }
    }
}

#[derive(derive_mmio::Mmio)]
#[repr(C)]
struct Uart {
    #[mmio(Inner)]
    array: [inner::UartBank; 2],
}

fn main() {
    let mut uart = Uart { array: [inner::UartBank::fake(), inner::UartBank::fake()] };

    // Safety: We're pointing at a real object
    let mut mmio_uart = unsafe { Uart::new_mmio(core::ptr::addr_of_mut!(uart)) };

    // we cannot safely access the array field without bounds checks

    let _inner_bank = mmio_uart.array_shared_unchecked(5);

    let _inner_bank = mmio_uart.array_unchecked(5);
}
