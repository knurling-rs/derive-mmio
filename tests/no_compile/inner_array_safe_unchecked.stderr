error[E0133]: call to unsafe function `MmioUart::<'_>::array_shared_unchecked` is unsafe and requires unsafe function or block
  --> tests/no_compile/inner_array_safe_unchecked.rs:32:23
   |
32 |     let _inner_bank = mmio_uart.array_shared_unchecked(5);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function `MmioUart::<'_>::array_unchecked` is unsafe and requires unsafe function or block
  --> tests/no_compile/inner_array_safe_unchecked.rs:34:23
   |
34 |     let _inner_bank = mmio_uart.array_unchecked(5);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior
